### PR TITLE
Beam-2171 Power function

### PR DESCRIPTION
**Pull request by app-tarush**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by app-tarush** [comment]
@takidau please review this PR.

**Comment by aaltay** [comment]
R: @xumingming

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+          .power(SqlFunctions.toLong(leftOp.getValue()), SqlFunctions.toLong(rightOp.getValue()))));
+
+    } else if (SqlTypeName.APPROX_TYPES.contains(leftOp.getOutputType()) || SqlTypeName.APPROX_TYPES
+        .contains(rightOp.getOutputType())) {
+
+      result = BeamSqlPrimitive.of(SqlTypeName.DOUBLE, SqlFunctions
+          .power(SqlFunctions.toDouble(leftOp.getValue()),
+              SqlFunctions.toDouble(rightOp.getValue())));
+
+    } else if (SqlTypeName.BIGINT.equals(leftOp.getOutputType()) && SqlTypeName.DECIMAL
```
Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java]: @takidau I will change the condition from : 

```java
else if (SqlTypeName.BIGINT.equals(leftOp.getOutputType()) && SqlTypeName.DECIMAL
```

to

```java
else if (SqlTypeName.INT_TYPES.equals(leftOp.getOutputType()) && SqlTypeName.DECIMAL
```

you are correct.

Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java]: Sounds good, thank you. Can you please also rebase onto DSL_SQL to pick up the ROUND PR that I just merged (as well as the SQL → Sql rename that went in)?

Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java]: @takidau sure 👍

Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java]: @takidau Changes made please review.

**Comment by app-tarush** [comment]
@takidau please review this PR #3336. changes made as suggested.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java [hunk 2]
```diff
@@ -278,6 +279,12 @@ static BeamSqlExpression buildExpression(RexNode rexNode) {
         case "SIGN":
           ret = new BeamSqlSignExpression(subExps);
           break;
+        case "POWER":
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java]: just to confirm: there are two function: `POW` and `POWER`?

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java [hunk 3]
```diff
@@ -135,4 +135,58 @@
...
+    operands.clear();
+    operands.add(BeamSqlPrimitive.of(SqlTypeName.DOUBLE, 1.1));
+    operands.add(BeamSqlPrimitive.of(SqlTypeName.BIGINT, 1L));
+    assertEquals(1.1, new BeamSqlPowerExpression(operands).evaluate(record).getValue());
+
+    // power(integer, double) => double
+    operands.clear();
+    operands.add(BeamSqlPrimitive.of(SqlTypeName.INTEGER, 2));
+    operands.add(BeamSqlPrimitive.of(SqlTypeName.DOUBLE, 2.2));
+    assertEquals(4.59479341998814,
```
Comment (xumingming) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java]: avoid usage of magic number.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java [hunk 4]
```diff
@@ -135,4 +135,58 @@
...
+        new BeamSqlPowerExpression(operands).evaluate(record).getValue());
+
+    // operand with a BeamSqlInputRefExpression
```
Comment (xumingming) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java]: Why you want to test a `BeamSqlInputRefExpression` here? seems not needed.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+public class BeamSqlPowerExpression extends BeamSqlMathBinaryExpression {
+
+  public BeamSqlPowerExpression(List<BeamSqlExpression> operands) {
+    super(operands);
+  }
+
+  @Override public BeamSqlPrimitive<? extends Number> calculate(BeamSqlPrimitive leftOp,
+      BeamSqlPrimitive rightOp) {
+    BeamSqlPrimitive result = null;
+    if (SqlTypeName.INT_TYPES.contains(leftOp.getOutputType()) && SqlTypeName.INT_TYPES
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java]: Do we simply return a `DOUBLE` just like the underlying `Math.pow` do?

**Comment by xumingming** [comment]
R: @xumingming

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java (hunk 2 above)
Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java]: @xumingming they both are similar.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java (hunk 3 above)
Comment (app-tarush) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java]: @xumingming I will do the change.

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java (hunk 4 above)
Comment (app-tarush) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java]: @xumingming it was a casual test, if wanted to be deleted then I can do that.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java (hunk 5 above)
Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java]: @xumingming I have done it in a way that is similar to mysql, earlier I was returning the type double but getting reviews on that so made it similar to mysql.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java (hunk 2 above)
Comment (takidau) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java]: Have both cases fall through to the same constructor, then?

File: dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java (hunk 4 above)
Comment (xumingming) [inline dsls/sql/src/test/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlMathBinaryExpressionTest.java]: from UnitTest point of view, it is not needed.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java (hunk 5 above)
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java]: I think just follow your previous PR, like `sin`, `asin` etc, just return a `DOUBLE`, it is simpler and consistent.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java (hunk 2 above)
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java]: I didn't found the `pow`(just find `power`) in calcite reference or calcite source code, can you point me where is `pow` defined? @app-tarush

Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/BeamSqlFnExecutor.java]: @xumingming I have to remove 'pow' as it was not mentioned in calcite.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java (hunk 5 above)
Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java]: @xumingming I also agree that behind the scenes it is getting changed into double.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import org.apache.calcite.runtime.SqlFunctions;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+/**
+ * {@code BeamSqlMathBinaryExpression} for 'POWER' function.
+ */
+public class BeamSqlPowerExpression extends BeamSqlMathBinaryExpression {
+
+  public BeamSqlPowerExpression(List<BeamSqlExpression> operands) {
+    super(operands);
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java]: Specify the outputType to `Double` .

**Comment by app-tarush** [comment]
@xumingming Please review the changes.

**Comment by xumingming** [comment]
outputType is not set to `Double`?

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java (hunk 6 above)
Comment (app-tarush) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/interpreter/operator/math/BeamSqlPowerExpression.java]: I have initialised through 
```java
this.outputType = SqlTypeName.DOUBLE;
```
you wanted me to do this through constructor

**Comment by takidau** [comment]
Merged. Didn't explode the test method, since there are already a number of other similar test methods in that file. We can clean those up later. Feel free to close, thanks!

**Comment by app-tarush** [comment]
Thanks 👍

**Pull request closed without merging**